New VMM extension integration mechanism

JaredWright opened the pull request:
1) New vmm_extension() macro for registering bareflank extensions to the
build system from build configuration file

2) New VMM extension-specific config variables and extension-specific macros

3) Added vmm_ex.cmake to be included by all Bareflank extensions

4) Updated build system documentation

[RFC] #545

Signed-off-by: JaredWright <jared.wright12@gmail.com>

cpavlina commented on scripts/docs/build_instructions.md:
````diff
@@ -0,0 +1,140 @@
+
+* Arch Linux:
+```
+sudo pacman -S git linux-headers nasm clang cmake
````
Also `base-devel`

rianquinn commented on scripts/cmake/config/bfconfig_template.cmake:
```diff
@@ -0,0 +1,41 @@
+# Bareflank Build Configurations
+# ------------------------------------------------------------------------------
+
+# set(BUILD_TYPE Debug)
```
We should provide documentation in the template so that all you have to do is read and modify the template.

JaredWright replied:
How much documentation should go in the template? I choose the commented out lines in the template as specific examples of 90% of what you would want to do with this file, namely:

* Setting a string type config
* Setting a boolean type config
* Adding an extension from GitHub
* Adding an extension from a local directory

What else should be added? Maybe a link or reference to the build instructions readme? One-sentence descriptions of what each section is for?

rianquinn replied:
I am a fan of redundancy, so in this case I would say adding the build instructions, and then a link to all of the options that are available in another doc would be fine so long as each option is documented somewhere.

JaredWright replied:
How about adding something like this:

```
# For instructions on configuring and building Bareflank, see hypervisor/scripts/docs/build_instructions.md

# To view all available configurations, use ccmake or cmake-gui from your build directory:
#      cd </path/to>/build; ccmake .
```

I've been avoiding creating a file that documents all available build configurations because those things are going to change, get added, and get dropped really fluidly as we continue development. As the build README indicates, the best way to view all of the available configs is to use ccmake or cmake-gui. From those, you can see every available option, the possible values for each, and a description for each. If we go the README route for build configs, we'll have to remember to add/modify two things every time we touch a build config.

rianquinn commented on scripts/cmake/config/default_ex.cmake:
```diff
@@ -0,0 +1,129 @@
+    DESCRIPTION "Path to the top directory for this VMM extension"
+)
+
+add_config(
```
Are these optional? If not they should be, and our example should be the simplest possible config with options commented out with comments on how to use them if desired. Most of these seem like things we should be able to "infer" based on our "preferred" directory layout

JaredWright replied:
The configs in this file are what get passed into the extension by the base build system, set to their default values. Extension writers can override these values in their extension-specific configuration files if desired. The configs here are exactly what you described: things we should be able to "infer" based on our "preferred" directory layout

JaredWright replied:
After I read your comment again, I think you're viewing this file as the "template" config file for an extension. That was not the intent here, this is meant to be setup by the base build system and "passed into" the extension. Extensions then define their own config files to define more values, or override the ones here

rianquinn replied:
ok yeah, so I think the issue for me was the "_ex", as I read that as example. Please rename to "_extension" and you should be good here

rianquinn commented on scripts/cmake/config/default_ex.cmake:
```diff
@@ -0,0 +1,129 @@
+    SKIP_VALIDATION
+)
+
+if(VMM_EX_IS_UNTTEST_BUILD)
```
Is there a way to simplify this? Everything else is pretty nice to look at, and seems easy to follow, this one part stands out as "yuk".

JaredWright replied:
I agree that this is a bit "yuk", but I can't think of a simpler way around it. The real problem that this addresses is that there seems to be a pattern present in every VMM component (including the base components) that most of the source code needs to be re-built during a unit test build, but you still need to the ability to exclude a few things from a unit test build. With that in mind, I thought extension writters should have a variable available to check if the build system is currently building the extension normally, or if it was building the unit tests for the extension.

In the old build system, this same check was all over the place, but written as if(VMM_TOOLCHAIN_FILE AND ENABLE_UNITTESTING). I thought that was more yuk that having this variable available. 

Any alternative suggestions?

rianquinn replied:
Your fine here as well since this is not something the user will be coding themselves.

JaredWright commented:
Another general comment I wanted to bring up about this PR:

I think that our build system is starting to cross over the line between normal-CMake and black-magic CMake. In other words, I just want to make sure that everyone is ok with the trade off of adding a non-trivial amount of complexity to our build system, with the payoff of ease-of-use for extension developers.

The only reason I bring this is that CMake has a steep learning curve for future maintenence, and we seem to be taking full advantage of all it has to offer.

JaredWright commented:
Most recent commit adds the following:

1. Added more instructions to the template bfconfig.cmake file
2. Renamed all the "*_ex" cmake files to "*_extension"
